lucet-runtime/lucet-runtime-internals/src/future.rs

```diff
@@ -150,91 +169,179 @@ impl InstanceHandle {
+    /// [`Instance::resume_with_val()`], may panic if the instance needs to block on an async function.
+    /// Use this function instead, which will resume the instance within an async context.
+    ///
+    /// The provided value will be dynamically typechecked against the type the guest expects to
+    /// receive, and if that check fails, this call will fail with `Error::InvalidArgument`.
+    ///
+    /// See [`Instance::resume()`], [`Instance::resume_with_val()`], and [`Instance::run_async()`].
+    ///
+    /// # Safety
+    ///
+    /// The foreign code safety caveat of [`Instance::run()`](struct.Instance.html#method.run)
+    /// applies.
+    pub fn resume_async<'a>(&'a mut self, val: Option<impl Any + 'static + Send>) -> RunAsync<'a> {
+        let val = match val {
+            Some(val) => Box::new(val) as Box<dyn Any + 'static + Send>,
+            None => Box::new(EmptyYieldVal),
+        };
+
+        RunAsync {
+            inst: self,
+            inst_count_bound: DEFAULT_INST_COUNT_BOUND,
+            state: RunAsyncState::ResumeYielded(val),
         }
+    }
 
-        // Store the ResumeVal here when we get it.
-        let mut resume_val: Option<ResumeVal> = None;
-        loop {
-            // Run the WebAssembly program
-            let run_result = if self.is_yielded() {
-                // A previous iteration of the loop stored the ResumeVal in
-                // `resume_val`, send it back to the guest ctx and continue
-                // running:
-                self.resume_with_val_impl(
-                    resume_val
-                        .take()
-                        .expect("is_yielded implies resume_value is some"),
-                    true,
-                    runtime_bound,
-                )
-            } else if self.is_bound_expired() {
-                self.resume_bounded(
-                    runtime_bound.expect("should have bound if guest had expired bound"),
-                )
-            } else {
+    /// Returns a `RunAsync` that will asynchronously execute the guest instnace.
+    fn run_async_internal<'a>(&'a mut self, func: FunctionHandle, args: &'a [Val]) -> RunAsync<'a> {
+        RunAsync {
+            inst: self,
+            inst_count_bound: DEFAULT_INST_COUNT_BOUND,
+            state: RunAsyncState::Start(func, args),
+        }
+    }
+
+    /// Returns a `RunAsync` that will immediately fail with the given error, without executing the guest instance.
+    fn run_async_failed<'a>(&'a mut self, err: Error) -> RunAsync<'a> {
+        RunAsync {
+            inst: self,
+            inst_count_bound: DEFAULT_INST_COUNT_BOUND,
+            state: RunAsyncState::Failed(err),
+        }
+    }
+}
+
+/// A future implementation that enables running a guest instance which can call async hostcalls.
+///
+/// Lucet hostcalls are synchronous `extern "C" fn` functions called from WebAssembly. In that
```

Comment:
A few clarity improvements: this doc comment focuses solely on async hostcalls, but this future implements all bounded-execution yielding, right? I think that it would be better to describe the timeslicing and async-passing mechanism here in a bit more detail to avoid missing forest for trees -- maybe something like:

"A Future implementation that enables asynchronous execution of bounded slices of WebAssembly, and of underlying async hostcalls that it invokes.

To execute WebAssembly code in a system that uses an async runtime with an executor loop, we instrument the machine code to periodically yield back to the caller. The on-stack frames are preserved by stack switching: the Wasm code runs on a separate stack.

This allows for yielding from the Wasm guest itself, but we also wish to support yielding from `async` hostcalls. The Lucet hostcall interface is synchronous at the lowest level (C-ABI functions are invoked by generated machine code). To pass a future returned by an async hostcall back to the top-level executor, we store state on the `vmctx`, switch stacks back to the caller, and then return the hostcall's future wrapped by this future back to the caller.

To define an async hostcall..."

Clarity improvement: "At the machine-code level, Lucet hostcalls are `extern "C" fn` functions that are called synchronously from WebAssembly. To "pass through" the future from a hostcall up through WebAssembly to the top-level executor loop, we need to adapt the hostcalls to pass the future through vmctx state, and adapt the `Instance` run method to return a future that returns this state."

Reply:
(On scanning over this again today I noticed that the second "Clarity improvement" was an earlier version of the comment text I drafted that I forgot to delete. I think the first quote above is probably better.)